# *Project Title: Predicting Customer Churn in a Telecommunications Company*


## Import the Necessary Libraries

In [31]:
import pandas as pd
import numpy as np

## Reads data from CSV file 

In [32]:
df = pd.read_csv('Telco-Customer-Churn-Dataset.csv')
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

## Explore the data structure

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


### Overview of the dataset
- **Number of rows:** 7,043 customers
- **Number of columns:** 21 attributes
- **Data type:** Mixed object, int64, float64
- **Missing values:** All columns have 7,043 non-null values

### Main attributes:
- **Personal information:** customerID, gender, SeniorCitizen, Partner, Dependents
- **Services:** tenure, PhoneService, MultipleLines, InternetService
- **Add-on services:** OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies
- **Contract:** Contract, PaperlessBilling, PaymentMethod
- **Finance:** MonthlyCharges, TotalCharges
- **Target variable:** Churn (Yes/No)

## Transform - Data Cleaning

In [34]:
# Drop duplicates
df = df.drop_duplicates()


In [35]:
# Discover unique values ​​and check data consistency
df.columns
for col in df.columns:
    print (col, df[col].unique())
    print ('-'*50)

customerID ['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
--------------------------------------------------
gender ['Female' 'Male']
--------------------------------------------------
SeniorCitizen [0 1]
--------------------------------------------------
Partner ['Yes' 'No']
--------------------------------------------------
Dependents ['No' 'Yes']
--------------------------------------------------
tenure [ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]
--------------------------------------------------
PhoneService ['No' 'Yes']
--------------------------------------------------
MultipleLines ['No phone service' 'No' 'Yes']
--------------------------------------------------
InternetService ['DSL' 'Fiber optic' 'No']
--------------------------------------------------
OnlineSec

In [36]:
# Automatically converts the data types of columns to the appropriate type
df = df.convert_dtypes()


In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   string 
 1   gender            7043 non-null   string 
 2   SeniorCitizen     7043 non-null   Int64  
 3   Partner           7043 non-null   string 
 4   Dependents        7043 non-null   string 
 5   tenure            7043 non-null   Int64  
 6   PhoneService      7043 non-null   string 
 7   MultipleLines     7043 non-null   string 
 8   InternetService   7043 non-null   string 
 9   OnlineSecurity    7043 non-null   string 
 10  OnlineBackup      7043 non-null   string 
 11  DeviceProtection  7043 non-null   string 
 12  TechSupport       7043 non-null   string 
 13  StreamingTV       7043 non-null   string 
 14  StreamingMovies   7043 non-null   string 
 15  Contract          7043 non-null   string 
 16  PaperlessBilling  7043 non-null   string 


The `TotalCharges` column has an incorrect data type format.

In [38]:
# Cast data to numeric type
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")
df["TotalCharges"] = df["TotalCharges"].astype("float64")

In [39]:
# Count number of missing values
df.isnull().sum(axis = 0)


customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

***Fill in the missing values with rule value***    
Condition: `tenure` = 0  
Action: fill in `TotalCharges` = 0

In [47]:

df.loc[df['tenure'] == 0, 'TotalCharges'] = 0

In [48]:
df['TotalCharges'].isnull().sum(axis = 0)

np.int64(0)

In [42]:
df.describe()

SeniorCitizen     tenure  MonthlyCharges  TotalCharges
count         7043.0     7043.0          7043.0   7043.000000
mean        0.162147  32.371149       64.761692   2279.734304
std         0.368612  24.559481       30.090047   2266.794470
min              0.0        0.0           18.25      0.000000
25%              0.0        9.0            35.5    398.550000
50%              0.0       29.0           70.35   1394.550000
75%              0.0       55.0           89.85   3786.600000
max              1.0       72.0          118.75   8684.800000

The features like `tenure`, `MonthlyCharges`, and `TotalCharges` are within reasonable values. However, when comparing the mean and median, there is a significant difference, especially in the `TotalCharges` column. To verify, I applied the **IQR (Interquartile Range)** method to determine outliers.

In [43]:
def find_outliers_iqr(col):
    Q1 = col.quantile(0.25)
    Q3 = col.quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = col[(col < lower_bound) | (col > upper_bound)]
    return outliers, lower_bound, upper_bound

In [44]:
for col in ["MonthlyCharges", "TotalCharges"]:
    outliers, lb, ub = find_outliers_iqr(df[col])
    print(f"--- {col} ---")
    print(f"Lower bound: {lb}, Upper bound: {ub}")
    print(f"Outliers:\n{outliers}\n")


--- MonthlyCharges ---
Lower bound: -46.02499999999999, Upper bound: 171.375
Outliers:
Series([], Name: MonthlyCharges, dtype: Float64)

--- TotalCharges ---
Lower bound: -4683.525, Upper bound: 8868.675
Outliers:
Series([], Name: TotalCharges, dtype: float64)



The results show that both `MonthlyCharges` and `TotalCharges` are within the allowable range, with no extreme outliers.

In [45]:
df.to_csv("telco_clean.csv", index=False)
